### ISSUE TITLE
website-template.tpl is out of date

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5380]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)| @thijstriemstra reported|
|-|-|
|Trac ID|trac#5380|
|Type|defect|
|Created|2011-11-12 16:26:30Z|
|Branch|https://github.com/twisted/twisted/tree/website-template-5380|

(from [#4693](https://github.com/twisted/twisted/issues/4693)) The release document looks wrong, since it does direct the release manager to use [doc/core/howto/template.tpl](doc/core/howto/template.tpl) to generate the website documentation. Instead, it should be directing the release manager to use [doc/core/howto/website-template.tpl](doc/core/howto/website-template.tpl).

In turn, it seems `website-template.tpl` has gotten somewhat out of date for a few reasons:

 - It is missing the version footer.
 - It is missing at sitemeter and google analytics (adding this would remove a couple gross manual steps from the release process).
 - It should have a link to the latest version of the docs, as was just recently added to the api docs pages.

Attachments:

* [website-template.tpl](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/70b/70bd9ca836e33ad03b80b0b95fa751e8614e877f/7b8d9665a53182b46db144a3d6578dcd0882cbf2.tpl) (2458 bytes) - added by thijs on 2011-11-26 16:27:20Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5380 5380
type__defect defect
reporter__thijs thijs
priority__normal normal
milestone__totally_automated_release_infrastructure totally automated release infrastructure
branch__branches_website_template_5380 branches/website-template-5380
branch_author__thijs thijs
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1321115190000000 1321115190000000
changetime__1328730632000000 1328730632000000
version__None None
owner__thijs thijs
cc__thijs cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_1"></a>|@thijstriemstra commented|
|-|-|

(In [33111]) Branching to 'website-template-5380'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_2"></a>|@thijstriemstra commented|
|-|-|

(In [33112]) Reference the website-template.tpl instead, refs [#5380](https://github.com/twisted/twisted/issues/5380)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

(In [33113]) add version link and site javascript, refs [#5380](https://github.com/twisted/twisted/issues/5380)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra commented|
|-|-|

(In [33114]) correct index link, add current link. refs [#5380](https://github.com/twisted/twisted/issues/5380)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra commented|
|-|-|

Replying to [thijs](#5380):
> (from [#4693](https://github.com/twisted/twisted/issues/4693)) The release document looks wrong, since it does direct the release manager to use [doc/core/howto/template.tpl](doc/core/howto/template.tpl) to generate the website documentation. Instead, it should be directing the release manager to use [doc/core/howto/website-template.tpl](doc/core/howto/website-template.tpl).

This has been fixed on ReleaseProcess. Also added a link to this ticket.

> 
> In turn, it seems template.tpl has gotten somewhat out of date for a few reasons:
> 

This should be `website-template.tpl`.

>  - It is missing the version footer.

Added.

>  - It is missing at sitemeter and google analytics (adding this would remove a couple gross manual steps from the release process).

Added. Still needs the find/sed fixes as described on ReleaseProcess.

>  - It should have a link to the latest version of the docs, as was just recently added to the api docs pages.

Added, but this also introduces a new find/sed line that needs to be run on the docs, I'll add it here for reference (needs to be added to the ReleaseProcess document when this ticket is merged to trunk).

```
find /tmp/twisted-release/doc -name '*.html' | xargs sed -i 's/container.style.display = &quot;&quot;/container.style.display = ""/'
```

Up for review.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess set owner to @thijstriemstra|
|-|-|

Thanks for working on this, thijs! A few comments:

 * urchin.js is from 2007, and it looks like there's a newer recommended way of adding Google Analytics tracking: http://www.google.com/support/googleanalytics/bin/answer.py?answer=174090. 
 * Site Meter similarly has a slightly updated version of its tracking javascript that doesn't use the WEBBOT tag: http://support.sitemeter.com/index.php?_m=knowledgebase&_a=viewarticle&kbarticleid=80

Can we use the newer versions of the Javascript for these tracking sites and update ReleaseProcess accordingly?

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra commented|
|-|-|

(In [33150]) Use latest google analytics and sitemeter scripts, refs [#5380](https://github.com/twisted/twisted/issues/5380)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Thanks for the review jesstess. I updated the scripts and also got rid of that additional `find` command from [comment:5 comment 5] by using single qoutes instead of double, this seems to work around [[#4544](https://github.com/twisted/twisted/issues/4544)](https://github.com/twisted/twisted/issues/4544) (so once this lands we can get rid of those analytics instructions on ReleaseProcess and move #4544 and [#4545](https://github.com/twisted/twisted/issues/4545) out of the [milestone:regular-releases] milestone. Would be good to have this in twisted 11.1 or the release manager will run into problems with the updated ReleaseProcess instructions.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra commented|
|-|-|

(In [33151]) Don't wait for sitemeter, refs [#5380](https://github.com/twisted/twisted/issues/5380)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra commented|
|-|-|

Marked [#5397](https://github.com/twisted/twisted/issues/5397) as duplicate of this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_13"></a>|@jesstess set owner to @thijstriemstra|
|-|-|

Sorry that it didn't make it into the release, but this looks great and ready to merge.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_14"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [33195]) Merge website-template-5380: Including version footer, a link to the latest version of the documentation,
and the latest analytics scripts for google and sitemeter in the website template for the howto documentation.

Author: thijs
Reviewer: jesstess
Fixes: [#5380](https://github.com/twisted/twisted/issues/5380)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `reopened`|
|-|-|

(In [33201]) Revert r33195 - test suite regression

The release automation tests in `twisted.python.test.test_release` fail with this
change due to the new name of the required template file.

Reopens: [#5380](https://github.com/twisted/twisted/issues/5380)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_16"></a>|@thijstriemstra set status to `new`|
|-|-|

`test_release` tests against `template.tpl`, but actual Twisted releases should use the updated `website-template.tpl` that includes the current link and analytics code.

exarkun mentioned that `website-template` belongs in website repository (lp:twisted-trac?) instead of the Twisted repository.

`bin/admin/build-tarballs` is hard-coded to use `template.tpl`, but this branch changed it to `website-template.tpl`, causing the test suite to fail.

So I suggest to add a template argument to the `bin/admin/build-tarballs`, if the argument is missing the default `template.tpl` will be used, included in the twisted source tree.

ReleaseProcess should be updated and explain how to obtain and use the `website-template.tpl` from the website repository.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_17"></a>|@thijstriemstra commented|
|-|-|

(In [33242]) Add 3rd parameter to build-tarballs, specifying the path to the custom (website) template. refs [#5380](https://github.com/twisted/twisted/issues/5380)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_18"></a>|@thijstriemstra set owner to @exarkun|
|-|-|

`website-template.tpl` has been deleted from the source tree and [[website-template.tpl](None/website-template.tpl) attached to this ticket]. Assigning it to exarkun so he can commit that file in the twisted website repository (I don't have commit rights there).

Release managers should now use this command to build the tarballs:

```
$ ./bin/admin/build-tarballs . /tmp/twisted-release /path/to/website-template.tpl
```

Without specifying a template the default [doc/core/howto/template.tpl](doc/core/howto/template.tpl) will be used.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

The website is lp:twisted-website.  Anyone can make a bzr branch and push it to launchpad, and I think you also have permission to push to the development focus branch.

  1. The `templatePath` API documentation is missing a word or something - _Path to the template file that used for the howto documentation._
  1. In `buildAllTarballs`, the new `isinstance` check seems weird.  A direct check against `None` would be better.  The entire check might be better if done in `DistributionBuilder._buildDocInDir` or `DistributionBuilder.build`.  Then again, since there is no code to actually _handle_ `TemplateDoesNotExist`, there doesn't seem to be much reason to make the check at all.  An unhandled `OSError` explaining that a file does not exist is about as user-friendly as an unhandled `TemplateDoesNotExist` saying the same thing.
  1. The new optional template path command line argument isn't very discoverable.  The usage errors that the script will emit  does not mention it.  This command may be complex enough to warrant the use of an `Options` subclass now.  At the very least, the usage information given in the error case of `BuildTarballsScript.main` needs to mention the new option.

Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_20"></a>|@thijstriemstra set status to `assigned`|
|-|-|

Replying to [exarkun](#note_19):
> The website is lp:twisted-website.  Anyone can make a bzr branch and push it to launchpad, and I think you also have permission to push to the development focus branch.

The [template file](http://bazaar.launchpad.net/~twisted-dev/twisted-website/trunk/view/head:/docs/website-template.tpl) has been added to the `docs` folder, I'll update the wiki accordingly.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_21"></a>|@thijstriemstra commented|
|-|-|

(In [33504]) Address review comments, refs [#5380](https://github.com/twisted/twisted/issues/5380)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_22"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Replying to [exarkun](#note_19):
>   1. The `templatePath` API documentation is missing a word or something - _Path to the template file that used for the howto documentation._
Fixed.

>   1. In `buildAllTarballs`, the new `isinstance` check seems weird.  A direct check against `None` would be better.  The entire check might be better if done in `DistributionBuilder._buildDocInDir` or `DistributionBuilder.build`.  Then again, since there is no code to actually _handle_ `TemplateDoesNotExist`, there doesn't seem to be much reason to make the check at all.  An unhandled `OSError` explaining that a file does not exist is about as user-friendly as an unhandled `TemplateDoesNotExist` saying the same thing.
Reverted it to OSError, eg.:
```
build-docs.py: /path/to/website-template.tpl: No such file or directory
```
>   1. The new optional template path command line argument isn't very discoverable.  The usage errors that the script will emit  does not mention it.  This command may be complex enough to warrant the use of an `Options` subclass now.  At the very least, the usage information given in the error case of `BuildTarballsScript.main` needs to mention the new option.

Added a description for the third optional argument.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

Thanks!

  1. Maybe the repeated `child` calls in `_release.py`, around line 985, could be replaced with use of the `descendant` method.
  1. The "re" module is imported in `release.py` but not used, maybe that could be removed.
  1. `BuildTarballsScript.main` is missing some test coverage.  The `len(args) == 2` case appears not to be exercised now.

I am now a little bit confused.  I tried to think about when someone would want to pass an alternate template path to `build-tarballs` but I couldn't figure anything out.  The facility certainly doesn't hurt, but it is superfluous with the current release process, right?  It's the yet nonexistent `build-docs` script that will really need this, I think?

Otherwise looks great.  I started a [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/website-template-5380).  If that looks good, and after addressing the above points, please merge.  Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_24"></a>|@thijstriemstra commented|
|-|-|

(In [33528]) address review comments, refs [#5380](https://github.com/twisted/twisted/issues/5380)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_25"></a>|@thijstriemstra set status to `assigned`|
|-|-|

Replying to [exarkun](#note_23):
> Thanks!
> 
>   1. Maybe the repeated `child` calls in `_release.py`, around line 985, could be replaced with use of the `descendant` method.
>   1. The "re" module is imported in `release.py` but not used, maybe that could be removed.
>   1. `BuildTarballsScript.main` is missing some test coverage.  The `len(args) == 2` case appears not to be exercised now.
> 

Fixed.

> I am now a little bit confused.  I tried to think about when someone would want to pass an alternate template path to `build-tarballs` but I couldn't figure anything out.  The facility certainly doesn't hurt, but it is superfluous with the current release process, right?  It's the yet nonexistent `build-docs` script that will really need this, I think?

Yeah, [ReleaseProcess#Buildhowtodocumentsforwebsite this script] is the temporary `build-docs` script. Once [#4500](https://github.com/twisted/twisted/issues/4500) lands we will have to rewrite this template stuff anyway, but nobody knows when that lore2sphinx is going to complete, so instead of adding this 'lore theme' to the trunk it now lives in lp:twisted-website as you suggested, where the sphinx theme will live as well I imagine.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_26"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [33529]) Merge website-template-5380: Moved outdated website template to the lp:twisted-website repository, used for building the howto documentation.

Author: thijs
Reviewer: jesstess, exarkun
Fixes: [#5380](https://github.com/twisted/twisted/issues/5380)
